src/apputils/clientsession.ts

```diff
@@ -0,0 +1,819 @@
+        return null;
+      }
+      if (result.accept) {
+        return kernel.restart().then(() => {
+          return kernel;
+        });
+      } else {
+        return kernel;
+      }
+    });
+  }
+
+  /**
+   * An interface for populating a kernel selector.
+   */
+  export
+  interface IKernelSearch {
+    /**
+     * The Kernel specs.
+     */
+    specs: Kernel.ISpecModels;
+
+    /**
+     * The current running sessions.
+     */
+    sessions?: IterableOrArrayLike<Session.IModel>;
+
+    /**
+     * The optional existing kernel model.
+     */
+    existing?: Kernel.IModel;
+
+    /**
+     * The preferred kernel name.
+     */
+    preferredName?: string;
+
+    /**
+     * The preferred kernel language.
+     */
+    preferredLanguage?: string;
+  }
+
+  /**
+   * Get the default kernel name given select options.
+   */
+  export
+  function getDefaultKernel(options: IKernelSearch): string {
+    return Private.getDefaultKernel(options);
+  }
+
+  /**
+   * Create a kernel dropdown list.
+   *
+   * @param options - The options used to populate the kernels.
+   *
+   * #### Notes
+   * Populates the list with separated sections:
+   *   - Kernels matching the preferred language (display names).
+   *   - "None" signifying no kernel.
+   *   - The remaining kernels.
+   *   - Sessions matching the preferred language (file names).
+   *   - The remaining sessions.
+   * If no preferred language is given or no kernels are found using
+   * the preferred language, the default kernel is used in the first
+   * section.  Kernels are sorted by display name.  Sessions display the
+   * base name of the file with an ellipsis overflow and a tooltip with
+   * the explicit session information.
+   */
+  export
+  function createKernelSelect(options: IKernelSearch): HTMLSelectElement {
+    return Private.createKernelSelect(options);
+  }
+}
+
+
+/**
+ * The namespace for module private data.
+ */
+namespace Private {
+  /**
+   * Select a kernel for the session.
+   */
+  export
+  function selectKernel(session: ClientSession, manager: Session.IManager): Promise<void> {
+    // Create the dialog body.
+    let body = document.createElement('div');
+    let text = document.createElement('label');
+    text.innerHTML = `Select kernel for: "${session.name}"`;
+    body.appendChild(text);
+
+    let options = getKernelSearch(session, manager);
+    let selector = ClientSession.createKernelSelect(options);
+    body.appendChild(selector);
+
+    let select = Dialog.okButton({ label: 'SELECT' });
+    return showDialog({
+      title: 'Select Kernel',
+      body,
+      buttons: [Dialog.cancelButton(), select]
+    }).then(result => {
+      // Change the kernel if a kernel was selected.
+      if (!result.accept) {
+        return;
+      }
+      let model = JSON.parse(selector.value) as Kernel.IModel | null;
+      if (!model) {
+        return session.shutdown();
+      }
+      return session.changeKernel(model).then(() => void 0);
```

Comment:
```typescript
return session.changeKernel(model).then(() => void 0);
```
Doesn't need the `.then(() => void 0);` because that already happens below.